### PR TITLE
fix: event API overloads

### PR DESCRIPTION
**Pull request by saikumarrs**
## PR Description

I've addressed all the outstanding issues with event API overloads.
Now, the types declaration files will properly show all the possible ways to invoke the APIs.

- `madge` dependency (to check circular references) has been removed to avoid vulnerability.
- Removed support for the following API usages:
  - `alias(callback)`
  - `alias(options, [callback])`
  - `group(callback)`
- Fixed a bug in the `page` API that would overwrite `name` and `category` fields if present in the `properties` argument for `page(properties)` usage.
- Updated the sanity suite to test for all the API overloads accordingly.
  - I've also fixed issues where some of the test cases in the suites were not completely independent of each other. The `reset` API is now called first before the actual API calls.
- Cleaned up Nx targets and size-limit configurations.
- I've also updated the NPM scripts to run only the affected projects during pre-commit and for the PRs.

## Linear task (optional)

https://linear.app/rudderstack/issue/SDK-2115/fix-issues-in-event-api-overloads-js-sdk

## Cross Browser Tests

Please confirm you have tested for the following browsers:

- [x] Chrome
- [x] Firefox
- [x] IE11

## Sanity Suite

- [x] All sanity suite test cases pass locally

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

## Summary by CodeRabbit

- **Bug Fixes**
	- Improved argument handling and cleanup for various analytics methods to prevent undefined values.
	
- **New Features**
	- Enhanced analytics methods (`alias`, `group`, `identify`, `page`, `track`) to support additional argument combinations, expanding their versatility.
	- Introduced new utility methods (`getAnonymousId`, `getUserId`, etc.) to retrieve specific analytical data points.

- **Tests**
	- Updated test cases to reflect changes in method usage and added new test cases for expanded scenarios.
	
- **Chores**
	- Enhanced CI workflows to improve security, code quality, and build size checks.
	- Adjusted CI scripts to handle environment variables dynamically based on the GitHub event context.

### CODE REVIEW

File: packages/analytics-js-common/src/utilities/eventMethodOverloads.ts [hunk 1]
```diff
@@ -231,7 +244,11 @@ const identifyArgumentsToCallOptions = (
     // in the Analytics class
     payload.userId = null;
     payload.traits = userId as Nullable<IdentifyTraits>;
-    payload.options = traits as Nullable<ApiOptions>;
+    if (!isFunction(traits)) {
+      payload.options = traits as Nullable<ApiOptions>;
+    } else {
+      delete payload.options;
+    }
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-common/src/utilities/eventMethodOverloads.ts:251]: How can delete operator impact performance?

[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-common/src/utilities/eventMethodOverloads.ts]: Addressed.

File: packages/analytics-js-common/src/utilities/eventMethodOverloads.ts [hunk 2]
```diff
@@ -103,11 +103,19 @@
...
   } else if (isObjectLiteralAndNotNull(name)) {
     delete payload.name;
     payload.properties = name as Nullable<ApiObject>;
-    payload.options = !isFunction(properties) ? (properties as Nullable<ApiOptions>) : null;
+    if (!isFunction(properties)) {
+      payload.options = properties as Nullable<ApiOptions>;
+    } else {
+      delete payload.options;
+    }
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-common/src/utilities/eventMethodOverloads.ts]: Addressed.

File: packages/analytics-js-common/src/utilities/eventMethodOverloads.ts [hunk 3]
```diff
@@ -103,11 +103,19 @@
...
     payload.properties = category as Nullable<ApiObject>;
-    payload.options = name as Nullable<ApiOptions>;
+    if (!isFunction(name)) {
+      payload.options = name as Nullable<ApiOptions>;
+    } else {
+      delete payload.options;
+    }
   } else if (isObjectLiteralAndNotNull(name)) {
     delete payload.name;
     payload.properties = name as Nullable<ApiObject>;
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-common/src/utilities/eventMethodOverloads.ts]: Addressed.

File: packages/analytics-js-common/src/utilities/eventMethodOverloads.ts [hunk 4]
```diff
@@ -103,11 +103,19 @@
     delete payload.name;
     delete payload.category;
     payload.properties = category as Nullable<ApiObject>;
-    payload.options = name as Nullable<ApiOptions>;
+    if (!isFunction(name)) {
+      payload.options = name as Nullable<ApiOptions>;
+    } else {
+      delete payload.options;
+    }
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-common/src/utilities/eventMethodOverloads.ts]: Addressed.

File: packages/analytics-js-common/src/utilities/eventMethodOverloads.ts [hunk 5]
```diff
@@ -363,19 +369,16 @@
...
     // Explicitly set null to prevent resetting the existing value
     // in the Analytics class
     payload.groupId = null;
     payload.traits = groupId as Nullable<ApiObject>;
-    payload.options = !isFunction(traits) ? (traits as Nullable<ApiOptions>) : null;
+    if (!isFunction(traits)) {
+      payload.options = traits as Nullable<ApiOptions>;
+    } else {
+      delete payload.options;
+    }
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-common/src/utilities/eventMethodOverloads.ts]: Addressed.

File: packages/analytics-js/__tests__/app/RudderAnalytics.test.ts [hunk 6]
```diff
@@ -162,7 +162,7 @@ describe('Core - Rudder Analytics Facade', () => {
     const groupSpy = jest.spyOn(analyticsInstance, 'group');
 
     rudderAnalytics.group(1234);
-    expect(groupSpy).toHaveBeenCalledWith({ groupId: '1234', traits: {} });
```
Comment (MoumitaM) [inline packages/analytics-js/__tests__/app/RudderAnalytics.test.ts:165]: Are we removing root level traits object from the payload for group call if not present?

File: packages/sanity-suite/src/testBookSuites/alias.js [hunk 7]
```diff
@@ -110,33 +116,115 @@ const aliasMethodSuite = {
...
       ],
       expectedResult: alias1ExpectedData,
-      triggerHandler: ['identify', 'alias'],
+      triggerHandler: ['reset', 'identify', 'alias'],
     },
     {
       id: 'alias2',
-      description: 'Call with to and from as arguments',
-      inputData: ['newCustomUserID', 'customUserID'],
+      description: 'rudderanalytics.alias(to, options)',
```
Comment (MoumitaM) [inline packages/sanity-suite/src/testBookSuites/alias.js:123]:
```suggestion
      description: 'rudderanalytics.alias(to, from)',
```

**Review by MoumitaM — CHANGES_REQUESTED** [review]

File: packages/analytics-js/__tests__/app/RudderAnalytics.test.ts (hunk 6 above)
Comment (saikumarrs) [reply packages/analytics-js/__tests__/app/RudderAnalytics.test.ts:165]: Yes. This is in line with the `identify` API implementation.

Comment (MoumitaM) [reply packages/analytics-js/__tests__/app/RudderAnalytics.test.ts:165]: If customers use this in transformation, that will break things for them.

File: packages/analytics-js-common/src/utilities/eventMethodOverloads.ts [hunk 8]
```diff
@@ -334,7 +342,7 @@
  * Normalise the overloaded arguments of the group call facade
  */
 const groupArgumentsToCallOptions = (
-  groupId: string | number | Nullable<ApiObject> | ApiCallback,
+  groupId: string | number | Nullable<ApiObject>,
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-common/src/utilities/eventMethodOverloads.ts:341]: Done.

File: packages/analytics-js-common/src/utilities/eventMethodOverloads.ts [hunk 9]
```diff
@@ -23,15 +23,15 @@
 };
 
 export type IdentifyCallOptions = {
-  userId?: string | null;
+  userId?: Nullable<string>;
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-common/src/utilities/eventMethodOverloads.ts:26]: Done.

File: packages/analytics-js/__tests__/app/RudderAnalytics.test.ts (hunk 6 above)
Comment (MoumitaM) [reply packages/analytics-js/__tests__/app/RudderAnalytics.test.ts:165]: Sounds good

**Review by MoumitaM — APPROVED** [review]

**Pull request merged**